### DISCUSSION TITLE
Where to add SMTP (and potentially later IMAP) libraries?

### DISCUSSION THREAD

**Discussion by CosmicToast** [janet-lang/janet#1103] [Polls]
I recently made [jurl](https://github.com/CosmicToast/jurl).
In its README, I mention that the low level native bindings could be used to build all sorts of other clients.
One such type of client you could build includes SMTP and IMAP.

I'm currently drafting out ways to get SMTP working (for my use-cases sending email is more important than receiving).
I'm wondering, would it be preferable to add these clients into jurl itself, or into a separate library?

The options:
* Add into jurl. Used via `(import jurl/smtp)` (for example).
* Add into a different package (e.g janet-email). Used via `(import email)` and `(import email/smtp)` and similar.

P.S. Since I'm currently workshopping the clients, if you have specific desires for what the API should be like, please let me know!
My current concept looks something like so:

```janet
# the address is extracted from the mention
(def client (smtp/new "smtps://my.smtp.server" "Chloe Kudryavtsev <me@smtp.me>" "my password"))
(:send client {:to ["Recipient One <one@example.com>"
                    "Recipient Two <two@example.com>"]
               :from "Chloe Kudryavtsev <me@smtp.me>" # optional
               :cc :me # :me implicitly inserts the username portion from above
               :bcc "Secret Person <secret@example.com>"} # BCC is stripped from the header list
              "body" # can also be a dictionary of mimetype -> body or a jurl/mime object
              :files ["upload.txt"]) # list of filenames or jurl/mime-compatible definitions, optional
```

**Comment by sogaiu** [comment]
May be starting with adding to Jurl and seeing how things go would be easier from a maintenance / development perspective?

**Comment by CosmicToast** [comment]
I should probably elaborate on the advantages and disadvantages (from the maintenance/development perspective).

In favor of Jurl:
* The email modules exclusively depend on curl (by way of jurl). All of the functionality is essentially janet wrapping on top of existing `jurl/native` code.
* People don't generally think of email when they see curl, and having it be in jurl helps people discover that it is a thing at all.

In favor of Janet-Email:
* People don't generally think of email when they see curl, so they might not realize jurl can be used for that.
* It's unlikely that the SMTP (and maybe later IMAP) modules will stabilize immediately, which means the (strict semver) version would likely need to be bumped quite a bit, while having its own module means it won't be affected.
* Having it be separate means the backend could be switched out (albeit it would be a breaking change).

In terms of maintenance and development there aren't any real concerns - I'm the only developer and maintainer (and likely will be for a while).

**Comment by pepe** [comment]
I would say look at spork 😂. And many thanks, I am going to need email soon 🥺.

**Comment by CosmicToast** [comment]
Now published (but not tagged, still have the README to modify and such).
`(import jurl/smtp)` then `(doc smtp)` and `(doc smtp/new)` (in that order).

**Reply by pepe** [reply]
Big love! I am checking that today.
